zkevm-circuits/src/evm_circuit/execution/call_staticcall.rs

```diff
@@ -0,0 +1,1283 @@
+use crate::{
+    evm_circuit::{
+        execution::ExecutionGadget,
+        param::{N_BYTES_ACCOUNT_ADDRESS, N_BYTES_GAS, N_BYTES_MEMORY_WORD_SIZE},
+        step::ExecutionState,
+        util::{
+            common_gadget::TransferGadget,
+            constraint_builder::{
+                ConstraintBuilder, ReversionInfo, StepStateTransition,
+                Transition::{Delta, To},
+            },
+            from_bytes,
+            math_gadget::{
+                BatchedIsZeroGadget, ConstantDivisionGadget, IsEqualGadget, IsZeroGadget,
+                MinMaxGadget,
+            },
+            memory_gadget::{MemoryAddressGadget, MemoryExpansionGadget},
+            select, sum, CachedRegion, Cell, Word,
+        },
+        witness::{Block, Call, ExecStep, Transaction},
+    },
+    table::{AccountFieldTag, CallContextFieldTag},
+    util::Expr,
+};
+use bus_mapping::evm::OpcodeId;
+use eth_types::{
+    evm_types::{GasCost, GAS_STIPEND_CALL_WITH_VALUE},
+    Field, ToLittleEndian, ToScalar,
+};
+use halo2_proofs::{circuit::Value, plonk::Error};
+use keccak256::EMPTY_HASH_LE;
+
+#[derive(Clone, Debug)]
+pub(crate) struct CallStaticCallGadget<F, const IS_CALL: bool> {
+    opcode: Cell<F>,
+    tx_id: Cell<F>,
+    reversion_info: ReversionInfo<F>,
+    current_address: Cell<F>,
+    is_static: Cell<F>,
+    depth: Cell<F>,
+    gas: Word<F>,
+    callee_address: Word<F>,
+    value: Option<Word<F>>,
+    is_success: Cell<F>,
+    gas_is_u64: IsZeroGadget<F>,
+    is_warm: Cell<F>,
+    is_warm_prev: Cell<F>,
+    callee_reversion_info: ReversionInfo<F>,
+    value_is_zero: IsZeroGadget<F>,
+    cd_address: MemoryAddressGadget<F>,
+    rd_address: MemoryAddressGadget<F>,
+    memory_expansion: MemoryExpansionGadget<F, 2, N_BYTES_MEMORY_WORD_SIZE>,
+    transfer: Option<TransferGadget<F>>,
+    callee_balance: Option<Cell<F>>,
+    callee_nonce: Cell<F>,
+    callee_code_hash: Cell<F>,
+    is_empty_nonce_and_balance: BatchedIsZeroGadget<F, 2>,
+    is_empty_code_hash: IsEqualGadget<F>,
+    one_64th_gas: ConstantDivisionGadget<F, N_BYTES_GAS>,
+    capped_callee_gas_left: MinMaxGadget<F, N_BYTES_GAS>,
+}
+
+impl<F: Field, const IS_CALL: bool> ExecutionGadget<F> for CallStaticCallGadget<F, IS_CALL> {
```

Comment:
seems it is not necessary to take use of `Const Generic ` to tell what current execution is call or staticcall . 

        let opcode = cb.query_cell();
          let is_call = PairSelectGadget::construct(
            cb,
            opcode.expr(),
            OpcodeId::CALL.expr(),
            OpcodeId::STATICCAll.expr(),
        );

Reply:
Fixed to calculate `is_call` in the gadget, and use it to check constraints.

But left a `Const Generic` parameter `HAS_VALUE` (with comments). Since both `call` and `callcode` pop 7 from stack, the third is `value`. But both `delegatecall` and `staticcall` only pop 6 from stack without `value`.

It seems that `cb.condition` (or `select::expr`) could not be used to handle different stack pops. The `stack_pop` is always run even if in `cb.condition`.

@DreamWuGit Do you have any better solution? WDYT? Thanks.

Reply:
Fix to increase `stack_pointer_offset` outside of `ConstraintBuilder` (since of stack pop `value` for opcode `CALL`).
So const generic `HAS_VALUE` is deleted and merge for both opcode `CALL` and `STATICCALL`.